### ISSUE TITLE
Understanding sly-mrepl--syntax-propertize

### ISSUE THREAD

**Issue by dieggsy** [joaotavora/sly#617]
I was playing around with adding syntax highlighting in `sly-mrepl-mode` and the following _almost_ got me there:
```emacs-lisp
(add-hook 'sly-mrepl-mode-hook
          (lambda ()
            (setq font-lock-defaults
                  '((lisp-cl-font-lock-keywords lisp-cl-font-lock-keywords-1 lisp-cl-font-lock-keywords-2)
                    nil t nil nil
                    (font-lock-mark-block-function . mark-defun)
                    (font-lock-extra-managed-props help-echo)
                    (font-lock-syntactic-face-function . lisp-font-lock-syntactic-face-function)))))
```

But I was seeing odd behavior - everything but the current input syntax highlighted as a comment. I found that the culprit was `sly-mrepl--syntax-propertize`, which I did a quick and dirty override for:

```emacs-lisp
(define-advice sly-mrepl--syntax-propertize (:override (beg end) d/no-comment-syntax)
  "Advice to avoid funky syntax stuff that sly-mrepl does for
unclear reasons. Maybe I will break everything down the line!
Exciting."
  nil)
```
This appears to do what I want, at least on the surface, but I'm concerned that there's a good reason not to obliterate this function. What is its purpose?

**Comment by joaotavora** [comment]
As far as I can remember, this function makes everything up to current prompt comment syntax (as the docstring says), to turn off parenthesis matching in those parts.  That's so that, when using things like `M-x electric-pair-mode` and other parenthesis matchers, they only consider the parenthesis of the current input prompt, and not any old output that may not having matching parenthesis.

**Comment by dieggsy** [comment]
Interesting. Maybe this should be applied at a per output level instead of on the whole buffer? I'd guess there's some better way to implement syntax highlighting to coexist with this, but I'm not too worried about overriding it for now.

**Comment by joaotavora** [comment]
It's not just the output, may also be some faulty input that doesn't balance parenthesis.

But you may argue that making the entire thing comment syntax is overkill . You could just tweak the syntax of those things
that are problematic.  But that's lot of work.  And comment syntax has other benefits.

What syntax highlighting exactly are planning to implement?

**Comment by dieggsy** [comment]
@joaotavora just common lisp syntax highlighting for input

**Comment by joaotavora** [comment]
Hmmm, so if it's just for the input, what's the trouble with making all the other non-input parts a certain syntax?

**Comment by dieggsy** [comment]
Well, but I'd like to preserve syntax highlighting on previous inputs as well

**Comment by joaotavora** [comment]
> Well, but I'd like to preserve syntax highlighting on previous inputs as well

In that case, I think you could "burn" the previously calculated properties with `font-lock-face` in those regions.  But I'd focus on getting font-lock to work correctly in input text first.

**Comment by monnier** [comment]
>     ;; matcher can be a function or a regex
>      (when ,(if (or (symbolp matcher) (functionp matcher))
>                 `(funcall ',matcher limit)
>               `(re-search-forward ,matcher limit t))

Side note: font-lock tests `stringp` instead (which is simpler and more
efficient).

>       ;; while the subexp can be anything, this search always can use zero
>       (let ((start (match-beginning 0))
>             (state (syntax-ppss)))
>         (and (or (eql 'sly-mrepl-input (get-text-property start 'field))
>                  (get-text-property start 'sly-mrepl--prompt)))))))

You don't use `state` here.


        Stefan

**Comment by dieggsy** [comment]
@stefan thanks for the notes! I did modify and iterate on [someone else's code](http://www.modernemacs.com/post/comint-highlighting/) and forgot i stopped using state (intentionally). I have cleaned up the code block in the original comment.

**Comment by joaotavora** [comment]
Another note:

```
(defun make-sly-font-lock-matcher (matcher)
  `(lambda (limit)
```

Seems a bit strange to be using backquoted `lambda`.  Are you sure you don't want to just use lexical binding?

**Comment by dieggsy** [comment]
Good call. I do in fact want to use lexical binding. Updated once more. Any thoughts on preserving comment/string highlighting in previous inputs? If there was a function that froze syntactic font locking as text properties or something like that...
